clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java

```diff
@@ -2412,15 +2412,36 @@ protected long measureLagWithCallToPubSub(
       PubSubTopic topic,
       int partition,
       long currentOffset) {
-    if (currentOffset < 0) {
+    return measureLagWithCallToPubSub(pubSubServerName, topic, partition, currentOffset, this::getTopicManager);
+  }
+
+  protected static long measureLagWithCallToPubSub(
+      String pubSubServerName,
+      PubSubTopic topic,
+      int partition,
+      long currentOffset,
+      Function<String, TopicManager> topicManagerProvider) {
+    if (currentOffset < OffsetRecord.LOWEST_OFFSET) {
+      // -1 is a valid offset, which means that nothing was consumed yet, but anything below that is invalid.
       return Long.MAX_VALUE;
     }
-    TopicManager tm = getTopicManager(pubSubServerName);
-    long endOffset = tm.getLatestOffsetCached(topic, partition) - 1;
+    TopicManager tm = topicManagerProvider.apply(pubSubServerName);
+    long endOffset = tm.getLatestOffsetCached(topic, partition);
     if (endOffset < 0) {
+      // A negative value means there was a problem in measuring the end offset, and therefore we return "infinite lag"
       return Long.MAX_VALUE;
+    } else if (endOffset == 0) {
```

Comment:
Why do we have this branch? I think it can be merged into below branch. EndOffset = 0 then currentOffset must be -1? then below formula also applies

Reply:
That is true. It was a suggestion from @sushantmane. It is not strictly necessary, but might be easier to read for some (maybe not all...?) people.

Reply:
After thinking about this more, I think maybe we should remove this branch.  Sorry for the trouble

Reply:
IMHO, I think it's fine to keep it, as it helps structure the comments to explain what each offset state means. Though the code could be one instruction shorter and achieve the same outcome, I don't think it matters in this case from a performance standpoint.